cli/asc.js

```diff
@@ -371,75 +357,105 @@ exports.main = function main(argv, options, callback) {
-            return path.join(_path, first, ascMain, second);
+        const actualDependee = path.resolve(dependee);
+        // paths = getPaths(path.join(baseDir, dependee));
+        let libStrippedSourcePathElements = sourcePath.replace(/^\~lib\/(.*)/, "$1")
+          .split(/\\|\//g);
+        let dependeeElements = actualDependee.split(SEP);
+
+        const getAscMainFolder = (absoluteModulePath) => {
+          if (SEP === "/" && !absoluteModulePath.startsWith("/")) {
+            // we are in a linux system, and we are dealing with nested packages
+            absoluteModulePath = "/" + absoluteModulePath;
           }
-          if (args.traceResolution) {
-            stderr.write("  in '" + realPath(sourcePath) + "'");
-          }
-          const plainName = sourcePath;
-          const indexName = sourcePath + "/index";
-          sourceText = readFile(realPath(plainName) + ".ts", baseDir);
-          if (sourceText !== null) {
-            sourcePath = plainName + ".ts";
-          } else {
-            sourceText = readFile(realPath(indexName) + ".ts", baseDir);
-            if (sourceText !== null) {
-              sourcePath = indexName + ".ts";
+          const packageJsonResult = readFile("package.json", absoluteModulePath);
+          if (packageJsonResult !== null) {
+            try {
+              const json = JSON.parse(packageJsonResult);
+              if (json.ascMain) {
+                return path.join(absoluteModulePath, path.dirname(json.ascMain));
+              }
+            } catch (ex) {
+              callback(Error("package.json in " + absoluteModulePath + " is malformed."));
             }
           }
-          if (sourceText !== null) {
-            if (args.traceResolution) {
-              stderr.write(EOL + "  -> '" + realPath(sourcePath) + "'" + EOL);
+          return path.join(absoluteModulePath, "assembly");
+        };
+        const moduleFolders = ["node_modules"].concat(args.path || []);
+        outer:
+        for (let i = 0, dependeeElementsLength = dependeeElements.length; i < dependeeElementsLength; i++) {
+          for (let j = 0, moduleFoldersLength = moduleFolders.length; j < moduleFoldersLength; j++) {
+            for (let k = 1, libElementsLength = libStrippedSourcePathElements.length; k <= libElementsLength; k++) {
+              const absoluteModuleFolder = dependeeElements.slice(0, dependeeElementsLength - i)
+                .concat(
+                  moduleFolders[j],
+                  libStrippedSourcePathElements.slice(0, k)
+                );
+              const ascMainFolder = getAscMainFolder(path.join(...absoluteModuleFolder));
+              const moduleSubFolders = libStrippedSourcePathElements.slice(k);
+              const moduleTargetFolder =
+                path.join(
+                  ascMainFolder,
+                  ...moduleSubFolders
+                );
+              let absolutePath = path.join(moduleTargetFolder, "index.ts");
```

Comment:
just use `path.resolve` instead `path.join` and you shouldn't require post-fixings next lines

Reply:
@MaxGraey Are you sure? The source of the folder name at that point comes from `absoluteModuleFolderPath` (in the latest commit) and if you call path.resolve(...["", "home", "etc"]) it doesn't return `"/home/etc"`

Looks like we can get away with this once by just checking to see if we need to append the `"/"` only once before all the checks start.

Reply:
```ts
> require("path").resolve("home")
'C:\\Users\\jtenner\\Desktop\\projects\\assemblyscript\\tests\\packages\\home'
```

Yeah, it won't return as expected. It will return relative from the current directory. I'm building an absolute module path and it needs to start with the fs root.

Reply:
you are right